apps/web/src/hooks/usePermitStatus.ts

```diff
@@ -0,0 +1,296 @@
+    if (!chainId) throw new Error('PERMIT: missing chainId')
+    if (!token) throw new Error('PERMIT: missing token')
+    if (!spender) throw new Error('PERMIT: missing spender')
+    if (!account) throw new Error('PERMIT: missing owner')
+    if (nonce === undefined) throw new Error('PERMIT: missing nonce')
+
+    const permit: Permit = generatePermitTypedData(token, nonce, spender)
+    const {
+      domain,
+      types,
+      values: message,
+    } = AllowanceTransfer.getPermitData(permit, getPermit2Address(chainId), chainId)
+
+    const signature = await signTypedDataAsync({
+      account,
+      domain,
+      primaryType: 'PermitSingle',
+      types,
+      message,
+    })
+
+    return {
+      ...permit,
+      signature,
+    }
+  }, [account, chainId, nonce, signTypedDataAsync, spender, token])
+}
+
+export enum PermitState {
+  IDLE,
+  REVOKING_ALLOWANCE,
+  APPROVING_ALLOWANCE,
+  APPROVING_PERMIT_AMOUNT,
+  DONE,
+}
+
+export const useAllowanceRequirements = (
+  approveByPermit2: boolean,
+  amount: CurrencyAmount<Token> | undefined,
+  spender?: Address,
+) => {
+  const { account } = useAccountActiveChain()
+  const { allowance: tokenAllowance } = useTokenAllowance(amount?.currency, account, spender)
+  const { allowance: permit2Allowance, permitAmount, expiration } = usePermitStatus(amount?.currency, account, spender)
+  const now = useCurrentBlockTimestamp() ?? 0n
+
+  const allowance = useMemo(
+    () => (approveByPermit2 ? permit2Allowance : tokenAllowance),
+    [approveByPermit2, permit2Allowance, tokenAllowance],
+  )
+
+  const requireRevoke = useMemo(() => {
+    const isMainnetUSDT =
+      amount?.currency?.chainId === ethereumTokens.usdt.chainId &&
+      isAddressEqual(amount?.currency.address, ethereumTokens.usdt.address)
+
+    if (!isMainnetUSDT) return false
+
+    return allowance?.greaterThan(0) && allowance.lessThan(amount)
+  }, [allowance, amount])
+
+  const requireApprove = useMemo(() => {
+    return amount && allowance?.lessThan(amount)
+  }, [allowance, amount])
+
+  const requirePermit = useMemo(() => {
+    return (approveByPermit2 && amount && permitAmount?.lessThan(amount)) || (Boolean(expiration) && now >= expiration)
+  }, [approveByPermit2, amount, permitAmount, expiration, now])
+
+  return useMemo(() => {
+    return {
+      requireApprove,
+      requireRevoke,
+      requirePermit,
+    }
+  }, [requireApprove, requirePermit, requireRevoke])
+}
+
+export const usePermitOrApprove = (
+  amount: CurrencyAmount<Currency> | undefined,
+  // if use approve, spender is the target address to grant allowance through erc20 SC
+  // if use permit2, spender is the target address to set permit amount through permit2 SC
+  spender?: Address,
+  onFinished?: () => Promise<SendTransactionResult | undefined>,
+  // only used for permit2
+  permit2signature?: Permit2Signature,
+  // only used for permit2
+  onUpdatePermit2Signature?: (signature: Permit2Signature) => void,
+) => {
```

Comment:
I'd suggest we design the function signature in a more intuitive way without callbacks.

Here's my idea for your reference
```typescript
const {
  // States
  permitAllowance, // token allowance (owner is the user) for permit2 contract
  permit2Allowance, // token allowance (owner is permit2 contract) for the signature
  permit2Signature, // the signature generated by signing the permit
  isApproving, // is approving the allowance for permit2
  isPermitting, // is signing permit
  isRevoking, // is revoking the allowance

  // Actions
  permit, // revoke approval first (if necessary), then approve to permit2 contract (if necessary) then permit
  revoke, // revoke the permit if possible, or just revoke the allowance for permit2 contract
} = usePermit2(amount, spender)
```

As for the approve state, I'd suggest we manage it in the component directly related to rendering. Because the state is meant to be reflected on the UI right? I think we can derive those UI states from the permit2 states described above. For example

```typescript
// This hook is dedicated for UI.
// E.g., if isRevoking then it's in the middle of the revoke action. 
// If none of the status is ongoing and permit2Signature is valid then it should be finished, etc.
const state = usePermit2State({
  isApproving,
  isRevoking,
  isPermitting,
  permit2Signature,
})
```

Reply:
Additionally, we may reuse the token approval if there's enough allowance (just an idea)
```typescript
function useTokenApproval(amount, spender) {
  const { isApproved } = usePermit2(amount, spender)
  const { allowance } = useTokenAllowance(amount, spender)

  // either allowance is enough, or is permitted
  const isApproved = ...
}
```

Reply:
implement in this PR #8992